protocol/network/4-architecture.md

```diff
@@ -0,0 +1,126 @@
+
+### 4.1.1 kad-routing
+
+kad-routing implements the Kademlia Routing table, where each peer holds a set of k-buckets, each of them containing several PeerInfo from other peers in the network. In order to find the whereabouts of these peers, it implements 3 discovery mechanisms:
+
+- mDNS-discovery
+- random-walk
+- bootstrap-list
+
+```
+┌────────────────────────────────────────────────────────────────┐
+│       Peer Routing                                             │
+│┌──────────────────────────────────────────────────────────────┐│
+││ kad-routing ┌──────────────┐┌──────────────┐┌──────────────┐ ││
+││             │mDNS-discovery││random-walk   ││bootstrap-list│ ││
+││             └──────────────┘└──────────────┘└──────────────┘ ││
+│└──────────────────────────────────────────────────────────────┘│
+│┌──────────────────────────────────────────────────────────────┐│
+││ mDNS-routing                                                 ││
+│└──────────────────────────────────────────────────────────────┘│
+│┌──────────────────────────────────────────────────────────────┐│
+││ other-routing-mechanisms                                     ││
+│└──────────────────────────────────────────────────────────────┘│
+└────────────────────────────────────────────────────────────────┘
+```
+
+#### 4.1.1.1 mDNS-discovery
+
+mDNS-discovery uses mDNS to emit beacons in the local area network to find if there are more peers available, these peers are extremely interesting because our link to them will have an low latency, plus if we are disconnected from the network, we will be leverage the content on those peers have cached.
+
+#### 4.1.1.2 random-walk
+
+Random walking consists on traversing the routing tables of the peers available in the network, learning about new peers along the way.
+
+#### 4.1.1.3 bootstrap-list
+
+A predefined set of peers available on the network in well known locations, so that a new created node can easily find more peers to connect in the network.
+
+### 4.1.2 mDNS-routing
+
+mDNS-routing uses mDNS probes to identify if local area network peers that have a given key or simply are present.
+
+
+
+
+
+
+
+
+## 4.2 Swarm
+
+### 4.2.1 Stream Muxer
+
+The stream muxer must implement the interface offered by [abstract-stream-muxer](https://github.com/diasdavid/abstract-stream-muxer).
+
+### 4.2.2 Protocol Muxer
+
+Protocol muxing is handled on the application level instead of the conventional way at the port level (where a different services/protocols listen at different ports). This enables us to support several protocols to be muxed in the same socket, saving the cost of doing NAT traversal for more than one port.
+
+Protocol multiplexing is done through [`multistream`](https://github.com/jbenet/multistream), a protocol to negoatiate different types of streams (protocols) using [`multicodec`](https://github.com/jbenet/multicodec)
```

Comment:
- @diasdavid: do you agree about `libp2p.Node` being the thing that does protocol muxing? i could be convinced otherwise, so we should discuss it if you feel strongly about it. I do think `libp2p.Node` should exist though, and NAT and Relay be part of _that_ instead of part of the `Swarm`.